File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/SetFollowCamProperties.java
package com.lumiyaviewer.lumiya.slproto.messages;

import com.google.common.primitives.UnsignedBytes;
import com.lumiyaviewer.lumiya.slproto.SLMessage;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.UUID;

public class SetFollowCamProperties extends SLMessage {
    public ArrayList<CameraProperty> CameraProperty_Fields = new ArrayList<>();
    public ObjectData ObjectData_Field;

    public static class CameraProperty {
        public int Type;
        public float Value;
    }

    public static class ObjectData {
        public UUID ObjectID;
    }

    public SetFollowCamProperties() {
        this.zeroCoded = false;
        this.ObjectData_Field = new ObjectData();
    }

    public int CalcPayloadSize() {
        return (this.CameraProperty_Fields.size() * 8) + 21;
    }

    public void Handle(SLMessageHandler sLMessageHandler) {
        sLMessageHandler.HandleSetFollowCamProperties(this);
    }

    public void PackPayload(ByteBuffer byteBuffer) {
        byteBuffer.putShort(-1);
        byteBuffer.put((byte) 0);
        byteBuffer.put((byte) -97);
        packUUID(byteBuffer, this.ObjectData_Field.ObjectID);
        byteBuffer.put((byte) this.CameraProperty_Fields.size());
        for (CameraProperty cameraProperty : this.CameraProperty_Fields) {
            packInt(byteBuffer, cameraProperty.Type);
            packFloat(byteBuffer, cameraProperty.Value);
        }
    }

    public void UnpackPayload(ByteBuffer byteBuffer) {
        this.ObjectData_Field.ObjectID = unpackUUID(byteBuffer);
        byte b = byteBuffer.get() & UnsignedBytes.MAX_VALUE;
        for (int i = 0; i < b; i++) {
            CameraProperty cameraProperty = new CameraProperty();
            cameraProperty.Type = unpackInt(byteBuffer);
            cameraProperty.Value = unpackFloat(byteBuffer);
            this.CameraProperty_Fields.add(cameraProperty);
        }
    }
}

File: Linkpoint/src/main/kotlin/com/linkpoint/slproto/messages/SetFollowCamProperties.kt
package com.linkpoint.slproto.messages

import com.google.common.primitives.UnsignedBytes
import com.linkpoint.slproto.SLMessage
import java.nio.ByteBuffer
import java.util.ArrayList
import java.util.UUID

class SetFollowCamProperties : SLMessage() {
    public ArrayList<CameraProperty> CameraProperty_Fields = ArrayList<>()
    public ObjectData ObjectData_Field

    @JvmStatic
    class CameraProperty {
        public Int Type
        public Float Value
    }

    @JvmStatic
    class ObjectData {
        public UUID ObjectID
    }

    public SetFollowCamProperties() {
        this.zeroCoded = false
        this.ObjectData_Field = ObjectData()
    }

    public fun CalcPayloadSize(): Int {
        return (this.CameraProperty_Fields.size() * 8) + 21
    }

    fun Handle(sLMessageHandler: SLMessageHandler) {
        sLMessageHandler.HandleSetFollowCamProperties(this)
    }

    fun PackPayload(byteBuffer: ByteBuffer) {
        byteBuffer.putShort(-1)
        byteBuffer.put((Byte) 0)
        byteBuffer.put((Byte) -97)
        packUUID(byteBuffer, this.ObjectData_Field.ObjectID)
        byteBuffer.put((Byte) this.CameraProperty_Fields.size())
        for (CameraProperty cameraProperty : this.CameraProperty_Fields) {
            packInt(byteBuffer, cameraProperty.Type)
            packFloat(byteBuffer, cameraProperty.Value)
        }
    }

    fun UnpackPayload(byteBuffer: ByteBuffer) {
        this.ObjectData_Field.ObjectID = unpackUUID(byteBuffer)
        val b: Byte = byteBuffer.get() & UnsignedBytes.MAX_VALUE
        for (Int i = 0; i < b; i++) {
            val cameraProperty: CameraProperty = CameraProperty()
            cameraProperty.Type = unpackInt(byteBuffer)
            cameraProperty.Value = unpackFloat(byteBuffer)
            this.CameraProperty_Fields.add(cameraProperty)
        }
    }
}
